bittide-instances/src/Project/Handle.hs

```diff
@@ -37,15 +40,45 @@ expectLine = expectLine' ""
         Stop (Error msg) -> do
           putStrLn s1
           assertFailure msg
+expectLine (Just (timeout, msg)) h f = do
+  initTime <- getCurrentTime
+  expectLine' initTime "" h f
+ where
+  timeout' = toRational timeout
+  expectLine' t0 s0 h' f' = do
+    now <- getCurrentTime
+    if toRational (diffUTCTime t0 now) >= timeout'
+      then errorIO $ "expect line timed out on: `" <> msg <> "`. previously read: \n" <> s0
+      else do
+        line <- hGetLine h
+        let
+          trimmed = trim line
+          s1 = s0 <> "\n" <> line
+          cont = expectLine' t0 s1 h' f'
+        if null trimmed
+          then cont
+          else case f trimmed of
+            Continue -> cont
+            Stop Ok -> pure ()
+            Stop (Error msg') -> do
+              rest <- readRemainingChars h'
+              putStrLn $ "handle buffer dump:\n" <> rest
+              errorIO $
+                "expect line failed on `"
+                  <> msg
+                  <> "`\nmessage: `"
+                  <> msg'
+                  <> "`\npreviously read:\n"
+                  <> s1
```

Comment:
Would it make sense to create a custom return type to differentiate between `Success`, `Timeout` and `Closed`?

Reply:
That I'm not sure of - I'm not entirely certain whether that info is actually useful to the postproc functions since I haven't read over them properly, just updated the call locations for the functions I changed.